clients/src/main/java/org/apache/kafka/common/memory/SimpleMemoryPool.java

```diff
@@ -0,0 +1,141 @@
+    protected final AtomicLong startOfNoMemPeriod = new AtomicLong();
+    protected volatile Sensor oomTimeSensor;
+
+    public SimpleMemoryPool(long sizeInBytes, int maxSingleAllocationBytes, boolean strict) {
+        if (sizeInBytes <= 0 || maxSingleAllocationBytes <= 0 || maxSingleAllocationBytes > sizeInBytes)
+            throw new IllegalArgumentException("must provide a positive size and max single allocation size smaller than size."
+                + "provided " + sizeInBytes + " and " + maxSingleAllocationBytes + " respectively");
+        this.sizeBytes = sizeInBytes;
+        this.strict = strict;
+        this.availableMemory = new AtomicLong(sizeInBytes);
+        this.maxSingleAllocationSize = maxSingleAllocationBytes;
+    }
+
+    @Override
+    public ByteBuffer tryAllocate(int sizeBytes) {
+        if (sizeBytes < 1)
+            throw new IllegalArgumentException("requested size " + sizeBytes + "<=0");
+        if (sizeBytes > maxSingleAllocationSize)
+            throw new IllegalArgumentException("requested size " + sizeBytes + " is larger than maxSingleAllocationSize " + maxSingleAllocationSize);
+
+        long available;
+        boolean success = false;
+        //in strict mode we will only allocate memory if we have at least the size required.
+        //in non-strict mode we will allocate memory if we have _any_ memory available (so available memory
+        //can dip into the negative and max allocated memory would be sizeBytes + maxSingleAllocationSize)
+        long threshold = strict ? sizeBytes : 1;
+        while ((available = availableMemory.get()) >= threshold) {
+            success = availableMemory.compareAndSet(available, available - sizeBytes);
+            if (success)
+                break;
+        }
+
+        if (success) {
+            maybeRecordEndOfDrySpell();
+        } else {
+            if (oomTimeSensor != null) {
+                startOfNoMemPeriod.compareAndSet(0, System.currentTimeMillis());
+            }
+            log.trace("refused to allocate buffer of size {}", sizeBytes);
+            return null;
+        }
+
+        ByteBuffer allocated = ByteBuffer.allocate(sizeBytes);
+        bufferToBeReturned(allocated);
+        return allocated;
+    }
+
+    @Override
+    public void release(ByteBuffer previouslyAllocated) {
+        if (previouslyAllocated == null)
+            throw new IllegalArgumentException("provided null buffer");
+
+        bufferToBeReleased(previouslyAllocated);
+        availableMemory.addAndGet(previouslyAllocated.capacity());
+        maybeRecordEndOfDrySpell();
+    }
+
+    @Override
+    public long size() {
+        return sizeBytes;
+    }
+
+    @Override
+    public long availableMemory() {
+        return availableMemory.get();
+    }
+
+    @Override
+    public boolean isOutOfMemory() {
+        return availableMemory.get() <= 0;
+    }
+
+    @Override
+    public void setOutOfMemoryPeriodSensor(Sensor oomPeriodSensor) {
```

Comment:
Could we just set oomPeriodSensor in the constructor and get rid of this method?